File: netsim/netsim.v4_R_det/R/distrib.R
rnorm_s <- function(n, mean=0, sd=1, chi)
{
	id <<- id + 1
	nome <- sprintf("normali_%d.txt", id)
	if (exists("salva") && get("salva") == T) {
		if (!file.exists(nome)) {
			stop("Il file '" + nome + "' non esiste!\n")
		}
		v <- scan(nome)
		cat(file=fp, sprintf("+++Letti %d valore/i per '%s' da '%s'\n", length(v), chi, nome))
	}
	else {
		v <- rnorm(n, mean, sd)
		write(file=nome, v, ncolumns=length(v))
	}
	return (v)
}

rlnorm_s <- function(n, mean=0, sd=1, chi)
{
	id <<- id + 1
	nome <- sprintf("log-normali_%d.txt", id)
	if (exists("salva") && get("salva") == T) {
		if (!file.exists(nome)) {
			stop("Il file '" + nome + "' non esiste!\n")
		}
		v <- scan(nome)
		cat(file=fp, sprintf("+++Letti %d valore/i per '%s' da '%s'\n", length(v), chi, nome))
	}
	else {
		v <- rlnorm(n, mean, sd)
		write(file=nome, v, ncolumns=length(v))
	}
	return (v)
}

runif_s <- function(n, min=0, max=1, chi)
{
	id <<- id + 1
	nome <- sprintf("unif_%d.txt", id)
	if (exists("salva") && get("salva") == T) {
		if (!file.exists(nome)) {
			stop("Il file '" + nome + "' non esiste!\n")
		}
		v <- scan(nome)
		cat(file=fp, sprintf("+++Letti %d valore/i per '%s' da '%s'\n", length(v), chi, nome))
	}
	else {
		v <- runif(n, min, max)
		write(file=nome, v, ncolumns=length(v))
	}
	return (v)
}

sample_s <- function(x, size, replace=FALSE, prob=NULL, chi="sampleB")
{
	id <<- id + 1
	nome <- sprintf("sample_%d", id)
	if (exists("salva") && get("salva") == T) {
		if (!file.exists(nome, ".txt")) {
			if (!file.exists(nome, "-vuoto.txt")) {
				stop("Il file n. " + id + "per sample non esiste!\n")
			}
			cat(file=fp, sprintf("+++Vettore nullo per '%s' da '%s'\n", chi, nome))
			ris <- integer(0)
		}
		ris <- scan(nome)
		cat(file=fp, sprintf("+++Letti %d valore/i per '%s' da '%s'\n", length(ris), chi, nome))
	}
	else {
		ris <- sample(x, size, replace, prob)
		if (length(ris) == 0) {
			nome <- sprintf("%s-vuoto.txt", nome)
			write(file=nome, ris)
		}
		else {
			nome <- sprintf("%s.txt", nome)
			write(file=nome, ris, ncolumns=length(ris))
		}
	}
	return (ris)
}

File: netsim/sim/distrib.c
#include <stdio.h>

#include "distrib.h"


VETTOREd *leggi_seq_d(VETTOREd *ris, char *buf, int n)
{
	int i = 1;
	char *tmp;
	double ris1;

	CREAv_d(ris, n);
	tmp = strtok(buf, " "); // primo numero
	do {
		sscanf(tmp, "%lg", &ris1);
		tmp = strtok(NULL, " ");
		ASSEGNAv_d(ris, i, ris1);
		i++;
	} while (tmp != NULL && i <= n);
	return ris;
}

VETTOREi *leggi_seq_i(VETTOREi *ris, char *buf, int n)
{
	int i = 1;
	char *tmp;
	int ris1;

	CREAv_i(ris, n);
	tmp = strtok(buf, " "); // primo numero
	do {
		sscanf(tmp, "%d", &ris1);
		tmp = strtok(NULL, " ");
		ASSEGNAv_i(ris, i, ris1);
		i++;
	} while (tmp != NULL && i <= n);
	return ris;
}

VETTOREd *rnorm_s(VETTOREd *ris, int n, double mean, double sd, const char *chi)
{
#ifdef DET
	char buf[MAX_RIGA], nome_mod[256];
	FILE *fp;

	id++;
	snprintf(nome_mod, 256, "normali_%d.txt", id);
	fp = fopen(nome_mod, "r");
	if (!fp) {
		Rprintf("Errore in lettura per il file '%s'!\n", nome_mod);
		error("");
	}
	fgets(buf, MAX_RIGA, fp);
	ris = leggi_seq_d(ris, buf, n);
	fprintf(fp_det, "+++Letti %d valore/i per '%s' da '%s'\n", LENGTHv_d(ris), chi, nome_mod);
	fclose(fp);
#else
	int i;

	for (i = 1; i <= n; i++)
		ASSEGNAv_d(ris, i, rnorm(mean, sd));
#endif

	return ris;
}

VETTOREd *rlnorm_s(VETTOREd *ris, int n, double mean, double sd, const char *chi)
{
#ifdef DET
	char buf[MAX_RIGA], nome_mod[256];
	FILE *fp;

	id++;
	snprintf(nome_mod, 256, "log-normali_%d.txt", id);
	fp = fopen(nome_mod, "r");
	if (!fp) {
		Rprintf("Errore in lettura per il file '%s'!\n", nome_mod);
		error("");
	}
	fgets(buf, MAX_RIGA, fp);
	ris = leggi_seq_d(ris, buf, n);
	fprintf(fp_det, "+++Letti %d valore/i per '%s' da '%s'\n", LENGTHv_d(ris), chi, nome_mod);
	fclose(fp);
#else
	int i;

	for (i = 1; i <= n; i++)
		ASSEGNAv_d(ris, i, rlnorm(mean, sd));
#endif

	return ris;
}

VETTOREd *runif_s(VETTOREd *ris, int n, double a, double b, const char *chi)
{
#ifdef DET
	char buf[MAX_RIGA], nome_mod[256];
	FILE *fp;

	id++;
	snprintf(nome_mod, 256, "unif_%d.txt", id);
	fp = fopen(nome_mod, "r");
	if (!fp) {
		Rprintf("Errore in lettura per il file '%s'!\n", nome_mod);
		error("");
	}
	fgets(buf, MAX_RIGA, fp);
	ris = leggi_seq_d(ris, buf, n);
	fprintf(fp_det, "+++Letti %d valore/i per '%s' da '%s'\n", LENGTHv_d(ris), chi, nome_mod);
	fclose(fp);
#else
	int i;

	for (i = 1; i <= n; i++)
		ASSEGNAv_d(ris, i, runif(a, b));
#endif

	return ris;
}

